Auto Implementing Generic Methods

haven1433 opened the pull request:
- update version to 1.0.1
- include xml documentation with System.Delegation
- don't include generated code in the source control. Use AutoImplement to generate during build.
- add implementations for generic methods.
- changed Indent() method to Scope property, since the fact that it's creating a scope is more important and the indent is just a side effect.

haven1433 committed:
added .xml output to System.Delegation. Added files for making a NuGet package from a release. Note that the first NuGet package (1.0.0) will include everything that was in the 1.0.0 release, and in the future the .nuspec needed for a nuget package will actually be included in that release.

haven1433 committed:
adding some documentation to System.Delegation and creating .xml for it. This is meant to be used
updating version numbers 1.0.0 -> 1.0.1
using AutoImplement to generate the Event-using class that we want for functional testing, to verify that the event works as expected.

haven1433 committed:
adding implementations for generic methods. For decorators/composites, this is basically just adding the list of generic parameters where it is needed. For Stubs, generic methods had to be implemented in a special way, because you cannot have generic lambda expressions. So for Stubs, every generic method Foo<> gets a matching ImplementFoo<>(Foo<>) method, that takes a delegate. Those delegates are added to a dictionary, and calling Foo<> looks for a matching dictionary entry to call. If it doesn't find one, just do nothing (return default). If you want to create a single method that implements arbitrary generics instead of adding specific implementations one at a time, you can wrap your stub with a decorator and put the general implementation in the decorator.

haven1433 committed:
change folders in .nuspec so I don't have to move files around when making a nuget release

haven1433 committed:
Update README.md

added information about the new features being added in this pull request

haven1433 committed:
the indent method did indent, but the more important thing is that it creates a scope. Changed from a method to a property and renamed to be more clear.

haven1433 committed:
Merge branch 'haven1433-dev' of https://github.com/haven1433/autoimplement into haven1433-dev

rtzoeller commented on AutoImplement/StringWriter.cs:
```diff
@@ -17,7 +17,7 @@ public class StringWriter {
       }
 
-      public IDisposable Indent() => new IndentationScope(this);
+      public IDisposable Scope => new IndentationScope(this);
```
Does this being a property cause any weird behavior? I would think that weird scoping would occur if a breakpoint was tripped and the `StringWriter` was expanded in Visual Studio.

Since it has side effects, it should probably be a method.

haven1433 replied:
I feel that `using (writer.Scope)` is clear enough for a user to understand what's going on. However, I did rename the class to make it more clear that this is specifically generating C# source files, making the renamed behavior a bit more obvious.

It does have side effects (indenting), but it being a property causes no weird behavior and helps the usage read easier. Otherwise, as a method, it would probably need to be called something like `CreateScope()` (verb/noun pattern). As a property that returns an IDisposable, I feel that this is clear enough.

haven1433 committed:
updating based on comments on the review.
Refactoring the code to make it make a bit more sense.
Also handling the leading 'I' on interfaces a little better, so it's less likely to do something weird if the leading 'I' isn't there.

rtzoeller commented on AutoImplement/Program.cs:
```diff
@@ -46,9 +46,9 @@ public static class Program {
             var mainName = type.Name.Split('`')[0];
             if (mainName.StartsWith("I")) mainName = mainName.Substring(1); // strip leading 'I' from interface name
```
Does the logic to remove a leading "I" when it makes sense need to be single sourced? It seems inconsistent between different call sites.

https://github.com/haven1433/autoimplement/blob/3a60df4da51337df10fcd0595b0122408a032d40/AutoImplement/ExtensionMethods.cs#L58

haven1433 replied:
Good point. I went ahead and single-sourced all the places that were trying to remove the leading 'I'.

While looking for other things that didn't make sense, I also moved the code that adds the using statements into the builders, to decrease how much the overall program needs to know about the code that's being generated.

haven1433 committed:
Moved knowledge of usings into the builders, since that makes more sense.
Single-sourced the code for removing the leading 'I' from interfaces.
Since Stubs include the System namespace for System.Type, clean up the generated code to take advantage of the System namespace for other classes.
  (EventHandler, Func, Action)

rtzoeller commented on AutoImplement/ExtensionMethods.cs:
```diff
@@ -50,23 +50,21 @@ public static string CreateCsName(this Type type, string currentNamespace)
-
-         return (mainName, genericInformation);
+      /// <summary>
+      /// Most interfaces start with a leading 'I' that we don't want to be part of teh child class name.
```
Spelling: 'teh'

haven1433 committed:
spelling

haven1433 committed:
fix indent level code: code was indenting 4x as much as it shoudl.

haven1433 committed:
lines in MemberMetadata too long to read. Breaking some of them up.

haven1433 committed:
making documentation on PropertyImplementation more clear. Removing unused tag from EventImplementation docs.

haven1433 committed:
Update README.md

Add section explaining that generic constraints are not respected in the current version.

haven1433 committed:
Merge branch 'haven1433-dev' of https://github.com/haven1433/autoimplement into haven1433-dev
merging changes for readme
